command/kv_rollback.go

```diff
@@ -0,0 +1,258 @@
+}
+
+func (c *KVRollbackCommand) AutocompleteFlags() complete.Flags {
+	return c.Flags().Completions()
+}
+
+func (c *KVRollbackCommand) Run(args []string) int {
+	f := c.Flags()
+
+	if err := f.Parse(args); err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	var version *int
+	f.Visit(func(fl *flag.Flag) {
+		if fl.Name == "version" {
+			version = &c.flagVersion
+		}
+	})
+
+	args = f.Args()
+
+	switch {
+	case len(args) != 1:
+		c.UI.Error(fmt.Sprintf("Invalid number of arguments (expected 1, got %d)", len(args)))
+		return 1
+	case version == nil:
+		c.UI.Error(fmt.Sprintf("Version flag must be specified"))
+		return 1
+	case c.flagVersion <= 0:
+		c.UI.Error(fmt.Sprintf("Invalid value %d for the version flag", c.flagVersion))
+		return 1
+	}
+
+	var err error
+	path := sanitizePath(args[0])
+
+	client, err := c.Client()
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 2
+	}
+
+	mountPath, v2, err := isKVv2(path, client)
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 2
+	}
+
+	if !v2 {
+		c.UI.Error(fmt.Sprintf("K/V engine mount must be version 2 for rollback support"))
+		return 2
+	}
+
+	path = addPrefixToVKVPath(path, mountPath, "data")
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 2
+	}
+
+	// First, do a read to get the current version for check-and-set
+	var meta map[string]interface{}
+	{
+		secret, err := kvReadRequest(client, path, nil)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error doing pre-read at %s: %s", path, err))
+			return 2
+		}
+
+		// Make sure a value already exists
+		if secret == nil || secret.Data == nil {
+			c.UI.Error(fmt.Sprintf("No value found at %s", path))
+			return 2
+		}
+
+		// Verify metadata found
+		rawMeta, ok := secret.Data["metadata"]
+		if !ok || rawMeta == nil {
+			c.UI.Error(fmt.Sprintf("No metadata found at %s; rollback only works on existing data", path))
+			return 2
+		}
+		meta, ok = rawMeta.(map[string]interface{})
+		if !ok {
+			c.UI.Error(fmt.Sprintf("Metadata found at %s is not the expected type (JSON object)", path))
+			return 2
+		}
+		if meta == nil {
+			c.UI.Error(fmt.Sprintf("No metadata found at %s; rollback only works on existing data", path))
+			return 2
+		}
+
+		// Verify current data found
+		rawData, ok := secret.Data["data"]
+		if !ok || rawData == nil {
+			c.UI.Error(fmt.Sprintf("No data found at %s; rollback only works on existing data", path))
+			return 2
+		}
+		data, ok := rawData.(map[string]interface{})
+		if !ok {
+			c.UI.Error(fmt.Sprintf("Data found at %s is not the expected type (JSON object)", path))
+			return 2
+		}
+		if data == nil {
+			c.UI.Error(fmt.Sprintf("No data found at %s; rollback only works on existing data", path))
+			return 2
+		}
+	}
+
+	casVersion := meta["version"]
+
+	// Set the version parameter
+	versionParam := map[string]string{
+		"version": fmt.Sprintf("%d", c.flagVersion),
+	}
+
+	// Now run it again and read the version we want to roll back to
+	var data map[string]interface{}
+	{
```

Comment:
never seen this syntax before, do it create a new lexical scope?

Reply:
Yes. Here it's just for logical separation so I could use the same mechanism but not worry about declaration vs assignment.

Reply:
thats a nice trick, the lazy version of a `func` :D

Reply:
There's enough difference that a function having to check parameters would have ended up as long, and quite possibly more complicated.